### ISSUE TITLE
limit total number of toasts

### ISSUE THREAD

**Issue by Kilian** [timolins/react-hot-toast#31]
Hey! Awesome library. Something that would be useful for me is an option to set the maximum number of visible toasts, that will automatically dismiss the oldest one when the maximum is reached.

I could do that with `useToaster` and a filter but then I would have to replicate everything else `<Toaster />` already does, so I would prefer not to.

**Comment by timolins** [comment]
That sounds like a reasonable improvement that shouldn't be too hard to implement. Consider this as on the road map.

**Comment by m-yasir** [comment]
Hey @timolins. I have been looking at some open source projects to contribute to, and this looks like one awesome project I am willing to work on. Would you mind having me look at this issue?

**Comment by iwgx** [comment]
I'm also find this useful, since show multiple toast in a time a bit not needed

Is there any way how to do this with current version? At least without using `useToaster`

**Comment by timolins** [comment]
You can achieve by using `useToasterStore()` & `useEffect()`. We can read the state of all toasts, and dismiss them if our limit is reached.

```js
const { toasts } = useToasterStore();

const TOAST_LIMIT = 3

useEffect(() => {
  toasts
    .filter((t) => t.visible) // Only consider visible toasts
    .filter((_, i) => i >= TOAST_LIMIT) // Is toast index over limit?
    .forEach((t) => toast.dismiss(t.id)); // Dismiss – Use toast.remove(t.id) for no exit animation
}, [toasts]);

```

Check out this [CodeSandbox example](https://codesandbox.io/s/limit-toast-amount-oe4pg?file=/src/App.js:243-590).

**Comment by timolins** [comment]
I think the proposed solution works well enough – no need to add this as native API IMO.

**Comment by weeebdev** [comment]
We wanted to extend it in a corp ui-kit. Could you add a possibility to change `TOAST_LIMIT` from `<Toaster/>` props? Or should I open a PR? https://github.com/timolins/react-hot-toast/blob/0d417d90d7d6c081c33af45eb5d71bd664fcdb1d/src/core/store.ts#L4

**Comment by timolins** [comment]
Looks like the demand for this, so I feel like it's time to add a proper API. Will look into this for the next release.

**Comment by narciso-dev** [comment]
Any news?

**Comment by LuisEgan** [comment]
In the meantime, I used @timolins `filter` solution in a hook so I don't have to write that `useEffect` everywhere:

`useToast.ts`
```
import { useEffect, useState } from "react";
import t, { useToasterStore } from "react-hot-toast";

const useToast = () => {
  const { toasts } = useToasterStore();

  const [toastLimit, setToastLimit] = useState<number>(3);

  useEffect(() => {
    toasts
      .filter((tt) => tt.visible)
      .filter((_, i) => i >= toastLimit)
      .forEach((tt) => t.dismiss(tt.id));
  }, [toasts]);

  const toast = {
    ...t,
    setLimit: (l: number) => {
      if (l !== toastLimit) {
        setToastLimit(l);
      }
    },
  };

  return { toast };
};

export default useToast;
```

Still, hopefully it'll be part of the native API soon

**Comment by brandonrbridges** [comment]
@LuisEgan Yeah, I've used a similar solution as I wanted to set a max limit on mobile devices. 

Hoping to have an official API soon.

**Comment by jeremiah-olisa** [comment]
I created a custom toaster component based on previous comments,

`
const CustomRHToaster = () => {
  const { toasts } = useToasterStore();
  
  const TOAST_LIMIT = 10;

  useEffect(() => {
    toasts
      .filter((_toast) => _toast.visible) // Only consider visible toasts
      .filter((_, i) => i >= TOAST_LIMIT) // Is toast index over limit?
      .forEach((_toast) => toast.dismiss(_toast.id)); // Dismiss – Use toast.remove(_toast.id) for no exit animation
  }, [toasts])

  return <Toaster position='top-right' />

}
`

**Comment by relativelyrehan** [comment]
Any Update guys on this ??

**Comment by AmirHmZz** [comment]
Any updates on this?

**Comment by belalahmad20** [comment]
There's kind of a bad hack using css - 

Give a containerClassName in Toaster - 

` 
<Toaster
...
containerClassName="toaster-wrapper"
...
 `
 
 And simple put this in your css - 
 
 `
.toaster-wrapper > div {
    display: none !important;
}
.toaster-wrapper > div:first-child {
    display: flex !important;
}
 `
 
and you can use nth-child concept to show the number of toast you want.

**Comment by Werthis** [comment]
Any updates?

**Comment by jeremiah-olisa** [comment]
You can create a custom hook to do that

**Comment by omateusamaral** [comment]
if your case its  showing the same error unlimited times, you can use the option "id" to prevent this.
https://react-hot-toast.com/docs/toast
example:
 ```
 toast.success('Copied to clipboard!', {
  id: 'clipboard',
});```

**Comment by AmirHmZz** [comment]
Any updates @timolins ?

**Comment by siyao-polarr** [comment]
> if your case its showing the same error unlimited times, you can use the option "id" to prevent this. https://react-hot-toast.com/docs/toast example:
> 
> ```
> toast.success('Copied to clipboard!', {
>  id: 'clipboard',
> });```
> ```

I tried this method since the other method with toaster store would make the animation very funky and old toasts not disappearing not immediately. This method is much nicer since it just replaces the content of the toast. However it since like the last toast would linger for a long time since the duration has been accumulating for the same id? This feels like a bug and shouldnt happen though. The correct behavior should reset the timer but not accumulating it?

**Comment by nyctonio** [comment]
.toaster-wrapper > div { display: none !important; } .toaster-wrapper > div:nth-child(-n+3) { display: flex !important; }

this works

**Comment by nyctonio** [comment]
> ` <Toaster ... containerClassName="toaster-wrapper" ...`

 <Toaster ... containerClassName="toaster-wrapper" ...

**Comment by NickNaskida** [comment]
Any updates on this? This is a very basic feature that shouldn't be hard to add to the library

**Comment by the-iter8** [comment]
It is just absurd that the issue has been opened since 2021 and there is still no new addition in the api.

**Comment by relativelyrehan** [comment]
I think it's time someone create a fix for this and create one PR

**Comment by Shielsy** [comment]
Any update on this? I've just switched to using toastr  but I may have to find an alternative if this is not fixed/added.

**Comment by pranavgoel29** [comment]
Please add an inbuilt method/option for this, I had to implement the whole logic my self. 🫠
For this basic feature. And please also add info() and warning() methods. 🙏🏻

**Comment by simon-schuster** [comment]
Any updates on this? I think the number limit would be an awesome feature. Also adding info() and warning() mentioned by 
@pranavgoel29 would improve this package a lot.

**Comment by softmarshmallow** [comment]
Hi. here's a simple wrapper component `ToasterWithMax`


```tsx
"use client";

import React, { useEffect } from "react";
import toast, { Toaster, useToasterStore } from "react-hot-toast";

function useMaxToasts(max: number) {
  const { toasts } = useToasterStore();

  useEffect(() => {
    toasts
      .filter((t) => t.visible) // Only consider visible toasts
      .filter((_, i) => i >= max) // Is toast index over limit?
      .forEach((t) => toast.dismiss(t.id)); // Dismiss – Use toast.remove(t.id) for no exit animation
  }, [toasts, max]);
}

export function ToasterWithMax({
  max = 10,
  ...props
}: React.ComponentProps<typeof Toaster> & {
  max?: number;
}) {
  useMaxToasts(max);

  return <Toaster {...props} />;
}
```

**Comment by Muhammadhassan1226** [comment]
const TOAST_LIMIT = 0;


  useEffect(() => {
    toasts
      .filter((t) => t.visible) // Only consider visible toasts
      .filter((_, i) => i >= TOAST_LIMIT) // Is toast index over limit?
      .forEach((t) => toast.dismiss(t.id)); // Dismiss – Use toast.remove(t.id) for no exit animation
  }, [toasts]);



I need only one toast at a time and this works for me in toast custom type

**Comment by johnsonfash** [comment]
<Toaster
        position="bottom-right"
        containerClassName="toaster-wrapper"
/>


CSS

.toaster-wrapper>div {
  display: none !important;
}

.toaster-wrapper>div:nth-child(-n+5) {
  display: flex !important;
}

//5 means 5-toasts max
